File: 198-house-robber/198-house-robber.java
class Solution {
    public int rob(int[] nums) {
        int n = nums.length;
        if(n < 2 )return nums[0];
        int[] dp = new int[n];
        dp[0] = nums[0]; dp[1] = nums[1]; 
        for(int i = 2; i < n;i++){
            int i2 = (i-2 < 0) ? Integer.MIN_VALUE : dp[i-2];
            int i3 = (i - 3 < 0) ? Integer.MIN_VALUE : dp[i-3];
            dp[i] = Math.max(i2, i3) + nums[i];
        }
        return Math.max(dp[n-2],dp[n-1]);
    }
}

File: 198-house-robber/198-house-robber.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        n = len(nums)
        @cache
        def helper(i):
            if i == 0:
                return nums[0]
            elif i == 1:
                return max(nums[0],nums[1])
            
            return max(helper(i-1),helper(i-2) + nums[i])
        
        return helper(n-1)
            